source/interact-data/transaction.txt

```diff
@@ -0,0 +1,283 @@
+ended. {+odm+} never triggers ``after_rollback`` outside of a transaction.
+
+.. _mongoid-txn-low-level:
+
+Lower Level API
+---------------
+
+When using the lower level API, you must create a session before
+starting a transaction. You can create a session by calling the
+``with_session()`` method on a model class or an instance of a model.
+
+Then, you can start a transaction by calling the ``start_transaction()``
+method on a session, as shown in the following code:
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-lower-lvl-api
+   :end-before: end-lower-lvl-api
+   :language: ruby
+   :dedent:
+
+You can specify a read concern, write concern or read
+preference when starting a transaction by passing options to the
+``start_transaction()`` method:
+
+.. code-block:: ruby
+
+   session.start_transaction(
+       read_concern: {level: :majority},
+       write_concern: {w: 3},
+       read: {mode: :primary}
+   )
+
+When using this API, you must manually commit or end the transaction.
+You can use the ``commit_transaction()`` and ``abort_transaction()``
+methods on the session instance to manage the transaction lifecycle:
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-commit-abort
+   :end-before: end-commit-abort
+   :language: ruby
+   :dedent:
+
+.. note::
+
+   If a session ends and includes an open transaction, the transaction is
+   automatically ended.
+
+You can retry the transaction commit if it fails,
+as shown in the following code: 
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-commit-retry
+   :end-before: end-commit-retry
+   :language: ruby
+   :dedent:
+
+Client Behavior
+~~~~~~~~~~~~~~~
+
+To perform operations within a transaction, operations must use the same
+client that the session was initiated on. By default, all operations
+are performed by using the default client.
+
+To explicitly use a different client, use the ``with()`` method: 
+
+.. literalinclude:: /includes/interact-data/transaction.rb
+   :start-after: start-other-client
+   :end-before: end-other-client
+   :language: ruby
+   :dedent:
+
+.. _mongoid-txn-session:
+
+Perform Data Operations in a Session
+------------------------------------
+
+You can use sessions in {+odm+} in a similar way that you can
+perform a transaction. You can call the ``with_session()`` method on a
+model class or on an instance of a model and perform some operations in
+a block. All operations in the block will be performed in the context of
+single session.
+
+The following limitations apply when using sessions:
+
+- You cannot share a session across threads. Sessions are not thread-safe.
+
+- You cannot nest sessions. You cannot call the ``with_session()``
+  method on a model class or a model instance within the block passed to
+  the ``with_session()`` method on another model class or model instance.
```

Comment:
S: This sentence is a bit hard to untangle, but breaking it up might be difficult. Could you maybe provide a code example of what not to do?